### DISCUSSION TITLE
middleware hides errors from background tasks

### DISCUSSION THREAD

**Discussion by varun-seth** [fastapi/fastapi#11828] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import BackgroundTasks, FastAPI, Request
import time

app = FastAPI()

@app.get("/")
def get_demo(background_tasks: BackgroundTasks):
    def task1():
        raise ValueError("Task 1 failed")

    background_tasks.add_task(task1)

    return {"status": "ok"}


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    start_time = time.time()
    response = await call_next(request)
    process_time = time.time() - start_time
    response.headers["X-Process-Time"] = str(process_time)
    return response

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app)
```

### Description

When the middleware is added, the errors from background tasks are not logged.

### Operating System

macOS

### FastAPI Version

0.111.0

### Pydantic Version

1.10.13

### Python Version

 3.11.6

**Comment by gustavosett** [comment]
When using `BackgroundTasks`, it's important to note that errors in background tasks don't affect the main request flow. This means they often pass silently unless explicitly logged or handled. In your case, the middleware is not necessarily "hiding" the errors, but since middleware wraps around the request/response lifecycle, it might not capture errors from background tasks which execute asynchronously and outside the request scope.

Try using `logging` in your background task to ensure has proper logging configured
```py
import logging

def task1():
    try:
        # Simulated error
        raise ValueError("Task 1 failed")
    except Exception as e:
        logging.error(f"Error in background task: {e}")
```

**Comment by YuriiMotov** [comment, marked as answer]
This is a Starlette issue and it was 8ntroduced in Starlette  0.29.0

Here is a Starlette app code that demonstrates it:

```
from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.routing import Route
from starlette.background import BackgroundTask
from starlette.responses import JSONResponse

class CustomHeaderMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        response = await call_next(request)
        return response

async def bg_task():
    raise ValueError("TEST")

def homepage(request):
    task = BackgroundTask(bg_task)
    message = {'status': 'Ok'}
    return JSONResponse(message, background=task)

routes = [
    Route('/', homepage),
]

middleware = [
    Middleware(CustomHeaderMiddleware)
]

app = Starlette(routes=routes, middleware=middleware)
```

**Reply by Kludex** [reply]
https://github.com/encode/starlette/issues/2625
